In [10]:
!pip install openai python-dotenv pandas


## Declarar mis funciones necesarias

In [1]:
import os
import json
import pandas as pd
from openai import OpenAI
from dotenv import load_dotenv

load_dotenv()
client = OpenAI(api_key=os.environ.get("OPENAI_API_KEY"))

reports_df = pd.read_excel("../../resources/listado-informes.xlsx")
progress_file = "prompt_iteration_progress.json"
output_file = "instances_generated.json"
reports_folder = "../../reports-pdf"


In [8]:

folder_path = reports_folder

file_list = []
for f in os.listdir(folder_path):
    name, ext = os.path.splitext(f)
    ext = ext[1:] if ext else ''
    file_list.append({'name': name, 'extension': ext})

df = pd.DataFrame(file_list)

with open('en_local.json', 'w') as f:
    df.to_json(f, orient='records')

In [ ]:

with open('/content/en_local.json') as f:
    local = json.load(f)
with open('/content/en_drive.json') as f:
    drive = json.load(f)

local_names = {item['name'] + '.' + item['extension'] for item in local}
drive_names = {item['name'] + '.' + item['extension'] for item in drive}
missing = sorted(local_names - drive_names)

print(missing)

ValueError: Trailing data

In [12]:

def load_progress():
    if os.path.exists(progress_file):
        with open(progress_file, "r", encoding="utf-8") as f:
            return json.load(f).get("current_row", 0)
    return 0

def save_progress(current_row):
    with open(progress_file, "w", encoding="utf-8") as f:
        json.dump({"current_row": current_row}, f)


In [13]:
def load_combined_prompt(ident, title):
    combined = ""
    for i in range(1, 4):
        path = f"prompts/basic_question{i}.txt"
        with open(path, "r", encoding="utf-8") as file:
            prompt = file.read().replace("{{report_name}}", f"{ident} {title}")
            combined += prompt + "\n\n"
    return combined.strip()


In [14]:
def find_report_file(ident):
    for ext in ['.pdf', '.csv']:
        path = os.path.join(reports_folder, ident + ext)
        if os.path.exists(path):
            return path
    return None


In [15]:
assistant = client.beta.assistants.create(
    model="gpt-4.1",
    tools=[{"type": "file_search",}],
    name="Asistente para informes",
    
)


In [16]:
def analyze_with_openai(prompt, filepath):
    file = client.files.create(file=open(filepath, "rb"), purpose="assistants")
    thread = client.beta.threads.create()

    client.beta.threads.messages.create(
        thread_id=thread.id,
        role="user",
        content=prompt,
        attachments=[{
            "file_id": file.id,
            "tools": [{"type": "file_search", }]
        }]
    )

    run = client.beta.threads.runs.create_and_poll(thread_id=thread.id, assistant_id=assistant.id, timeout=1000)
    if run.status != "completed":
        raise Exception(f"Ejecución fallida: {run.status}")

    messages = client.beta.threads.messages.list(thread_id=thread.id)
    return messages.data[0].content[0].text.value


In [17]:
import re
import json

def extract_json_block(text):
    """
    Extrae el primer bloque JSON válido (que empieza con '[' y termina con ']')
    del texto proporcionado.
    """
    try:
        match = re.search(r'\[\s*{.*?}\s*\]', text, re.DOTALL)
        if match:
            json_str = match.group(0)
            return json.loads(json_str)
        else:
            print("⚠️ No se encontró un bloque JSON en la respuesta.")
            return None
    except Exception as e:
        print("❌ Error al parsear el JSON:", e)
        return None


In [18]:
def append_instances(new_text):
    all_instances = []
    if os.path.exists(output_file):
        with open(output_file, "r", encoding="utf-8") as f:
            all_instances = json.load(f)
    try:
        parsed = extract_json_block(new_text)
        if not parsed:
            print("⚠️ No se pudo extraer el JSON.")
            return
        all_instances.extend(parsed)
        with open(output_file, "w", encoding="utf-8") as f:
            json.dump(all_instances, f, ensure_ascii=False, indent=2)
    except Exception as e:
        print("❌ Error al guardar JSON:", e)


## Generar instancias por cada archivo con la api de OpenAI

In [19]:
start_row = load_progress()
total_rows = len(reports_df)

print(f"▶️ Empezando desde la fila {start_row} de {total_rows}")

for idx in range(start_row, total_rows):
    row = reports_df.iloc[idx]
    ident, title = row["ident"], row["title"]
    print(f"\n📄 Procesando ({idx+1}/{total_rows}) → {ident} - {title}")

    file_path = find_report_file(ident)
    if not file_path:
        print("⚠️ Archivo no encontrado, se salta.")
        save_progress(idx + 1)
        continue

    try:
        prompt = (
            "Te voy a dar 3 prompts para que me generes 5 instancias (2 prompts generan de una instancia, "
            "1 prompt genera 3)  y metes las instancias en un texto formato json.Dame las 5 instancias en un solo array json. NO ME DES código python o algo así. "
            "Dame directamente el texto.\n\n" + load_combined_prompt(ident, title)
        )
        response = analyze_with_openai(prompt, file_path)
        print(response)
        append_instances(response)
        print("✅ Instancias generadas y guardadas.")
    except Exception as e:
        print(f"❌ Error procesando {ident}: {e}")

    save_progress(idx + 1)

print("🎯 Proceso completado.")


▶️ Empezando desde la fila 664 de 715

📄 Procesando (665/715) → 748-CI-00648 - El daño que nos hacen: glifosato y guerra en Caquetá
[
  {
    "question": "¿Cuáles son los hallazgos principales del informe 748-CI-00648 El daño que nos hacen: glifosato y guerra en Caquetá sobre el conflicto armado colombiano?",
    "context": "El informe 748-CI-00648 documenta cómo la fumigación con glifosato, en el contexto de la política de lucha contra las drogas ilícitas, ha causado graves afectaciones a las comunidades rurales de Caquetá. Se evidencian desplazamientos forzados, daños a la salud y el ambiente, y la criminalización de pequeños cultivadores. Además, se subraya la persistencia de actores armados ilegales y la insuficiencia de la respuesta estatal para proteger a la población civil del conflicto armado, lo cual ha agravado las violaciones a los derechos humanos y la situación humanitaria de la región.",
    "answer": "El informe destaca que el uso de glifosato en el marco del conflicto a

## extraer las instancias que se imprimieron en consola

In [ ]:
import json
import re

def extract_json_arrays(text):
    json_arrays = []
    stack = []
    start = None
    in_string = False
    escape = False

    for i, char in enumerate(text):
        if char == '"' and not escape:
            in_string = not in_string
        if char == '\\' and not escape:
            escape = True
            continue
        escape = False

        if in_string:
            continue

        if char == '[':
            if not stack:
                start = i
            stack.append(char)
        elif char == ']':
            if stack:
                stack.pop()
                if not stack and start is not None:
                    fragment = text[start:i + 1]
                    try:
                        parsed = json.loads(fragment)
                        if isinstance(parsed, list):
                            json_arrays.append(parsed)
                    except:
                        pass
                    start = None

    return json_arrays

def main():
    with open('response.txt', 'r', encoding='utf-8') as f:
        content = f.read()

    arrays = extract_json_arrays(content)
    concatenated = []
    for arr in arrays:
        concatenated.extend(arr)

    with open('resultado_concatenado.json', 'w', encoding='utf-8') as f:
        json.dump(concatenated, f, ensure_ascii=False, indent=2)

    print(f"Total de elementos concatenados: {len(concatenated)}")

if __name__ == "__main__":
    main()


'merged_responses.json'

In [26]:
## load the json file
import json
import pandas as pd
import os
from pathlib import Path


# Load the JSON file
data={}
with open("resultado_concatenado.json", "r", encoding="utf-8") as file:
    data = json.load(file)

In [27]:
# filter data array by "context" key
filtered_data = [entry for entry in data if "context" in entry and entry["context"]!= ""]
print(len(filtered_data))


# Agregar el campo "generated_by" a cada objeto
new_data = []
for item in filtered_data:
    if not "answer" in item:
        print("⚠️ No se encontró la clave 'answer' en el objeto:", item)
        continue
    new_item = {
        "context": item["context"],
        "question": item["question"],
        "answer": item["answer"],
        "generated_by": "gpt-4.1",
    }
    new_data.append(new_item)
    

# Guardar el nuevo archivo
with open("instances_generated_enriched.json", "w", encoding="utf-8") as f:
    json.dump(new_data, f, ensure_ascii=False, indent=2)

print("Archivo enriquecido guardado como data_enriched.json")

71
Archivo enriquecido guardado como data_enriched.json


## Reintentar con los que no se generaron en el primer intento

In [24]:
start_row = 125
reports_df2 = pd.read_excel("failed_questions.xlsx")
total_rows = len(reports_df2)

print(f"▶️ Empezando desde la fila {start_row} de {total_rows}")

for idx in range(start_row, total_rows):
    row = reports_df2.iloc[idx]
    ident, title, questions= row["ident"], row["title"], row["questions"]
    print(f"\n📄 Procesando ({idx+1}/{total_rows}) → {ident} - {title}")

    file_path = find_report_file(ident)
    if not file_path:
        print("⚠️ Archivo no encontrado, se salta.")
        save_progress(idx + 1)
        continue

    try:
        prompt = (
            "TE voy a dar unas preguntas para que me generes instancias de un dataset en formato json. de acuerdo al archivo "+title+" que te voy a dar. "
            "el formato de las instancias que debes darme es el siguiente:\n\n" 
            "context: question: answer. es decir. te estoy dando las questions. tu dame lo demas. A continuacion las questions:\n\n" + questions
        )
        
        response =analyze_with_openai(prompt, file_path)
        print(response)
        append_instances(response)
        print("✅ Instancias generadas y guardadas.")
    except Exception as e:
        print(f"❌ Error procesando {ident}: {e}")

    save_progress(idx + 1)

print("🎯 Proceso completado.")

▶️ Empezando desde la fila 125 de 199

📄 Procesando (126/199) → 262-CI-00780 - 262-CI-00780
Aquí tienes una instancia del dataset en formato JSON, respondiendo a tu pregunta basada específicamente en la entrevista de la Red Mariposas de Alas Nuevas Construyendo Futuro sobre las causas estructurales del conflicto armado y sus recomendaciones para la no repetición, según el archivo "262-CI-00780":

```json
{
  "context": "Entrevista Caso #1 realizada por la Red Mariposas de alas Nuevas Construyendo Futuro a una mujer víctima del conflicto armado en el Cauca, Colombia, relatando experiencias de violencia física, sexual, racismo, desplazamiento forzado, y desatención institucional tanto por parte de actores armados (guerrilla) como instituciones del Estado.",
  "question": "¿Qué dice este informe 262-CI-00780 Caso #1: entrevista de la Red Mariposas de alas Nuevas Construyendo Futuro sobre las causas estructurales del conflicto armado y sus recomendaciones para la no repetición?",
  "answer

## Guardar en huggingface

In [1]:
from datasets import Dataset
import pandas as pd

# Cargar el nuevo JSON enriquecido como DataFrame
df = pd.read_json("dataset/test_instances_generated_v1_2.json")

# Convertir a Hugging Face Dataset
dataset = Dataset.from_pandas(df)

# Subir al Hub (sobrescribirá el dataset anterior, que ya estaba vacío)
dataset.push_to_hub("jdavit/colombian-conflict-SQA",split="test")

print("Nuevas instancias subidas exitosamente a jdavit/colombian-conflict-SQA")

c:\Users\Usuario\Desktop\univalle\tesis\Tesis\notebooks\utils\.conda\Lib\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm
Uploading the dataset shards: 100%|██████████| 1/1 [00:01<00:00,  1.36s/it]
c:\Users\Usuario\Desktop\univalle\tesis\Tesis\notebooks\utils\.conda\Lib\site-packages\huggingface_hub\file_download.py:144: UserWarning: `huggingface_hub` cache-system uses symlinks by default to efficiently store duplicated files but your machine does not support them in C:\Users\Usuario\.cache\huggingface\hub\datasets--jdavit--colombian-conflict-SQA. Caching files will still work but in a degraded version that might require more space on your disk. This warning can be disabled by setting the `HF_HUB_DISABLE_SYMLINKS_WARNING` environment variable. For more details, see https://huggingface.co/docs/huggingface_hub/how-to-cac

Nuevas instancias subidas exitosamente a jdavit/colombian-conflict-SQA


## Buscar el archivo con mas texto para saber cual es el mas largo y tenerlo como referencia

- el más grande tiene 162.000 palabras

Buscando los 10 archivos PDF/CSV más grandes...

Procesando los archivos...

Resultados:

Resultados:
| #  | Archivo            | Tipo | Tamaño (MB) | Palabras |
|----|--------------------|------|-------------|----------|
| 0  | 365-CI-01208.pdf   | PDF  | 415.01      | 27571    |
| 1  | 119-CI-00045.pdf   | PDF  | 358.58      | 339421   |
| 2  | 748-CI-00863.pdf   | PDF  | 327.56      | 91304    |
| 3  | 748-CI-00864.pdf   | PDF  | 327.56      | 91304    |
| 4  | 365-CI-01260.pdf   | PDF  | 270.28      | 219309   |
| 5  | 119-CI-00315.pdf   | PDF  | 258.05      | 0        |
| 6  | 1308-CI-02024.pdf  | PDF  | 210.18      | 168609   |
| 7  | 365-CI-01192.pdf   | PDF  | 169.5       | 154781   |
| 8  | 365-CI-01242.pdf   | PDF  | 152.88      | 103932   |
| 9  | 748-CI-00872.pdf   | PDF  | 142.92      | 87690    |

In [ ]:
import os
import json
from PyPDF2 import PdfReader
from tabulate import tabulate

def get_largest_files(directory, extensions=('.pdf', '.csv'), top_n=10):
    """Obtiene los archivos más grandes con extensiones específicas"""
    files = []
    for entry in os.scandir(directory):
        if entry.is_file() and entry.name.lower().endswith(extensions):
            files.append((entry.path, entry.stat().st_size))
    
    # Ordenar por tamaño (de mayor a menor) y tomar los top_n
    files.sort(key=lambda x: x[1], reverse=True)
    return [file[0] for file in files[:top_n]]

def count_words_pdf(file_path):
    """Cuenta palabras en un archivo PDF"""
    try:
        reader = PdfReader(file_path)
        text = ""
        for page in reader.pages:
            text += page.extract_text() or ""
        return len(text.split())
    except Exception as e:
        print(f"Error procesando PDF {file_path}: {str(e)}")
        return 0

def count_words_csv(file_path):
    """Cuenta palabras en un archivo CSV"""
    try:
        with open(file_path, 'r', encoding='utf-8') as f:
            content = f.read()
            return len(content.split())
    except Exception as e:
        print(f"Error procesando CSV {file_path}: {str(e)}")
        return 0

def process_files(file_paths):
    """Procesa los archivos y devuelve los resultados"""
    results = []
    for file_path in file_paths:
        size_mb = os.path.getsize(file_path) / (1024 * 1024)
        if file_path.lower().endswith('.pdf'):
            word_count = count_words_pdf(file_path)
        else:
            word_count = count_words_csv(file_path)
        
        results.append({
            "Archivo": os.path.basename(file_path),
            "Tipo": "PDF" if file_path.lower().endswith('.pdf') else "CSV",
            "Tamaño (MB)": f"{size_mb:.2f}",
            "Palabras": word_count
        })
    return results

def main():
    directory = input("Ingrese la ruta del directorio a analizar (deje vacío para usar el actual): ").strip()
    if not directory:
        directory = os.getcwd()
    
    if not os.path.isdir(directory):
        print("El directorio especificado no existe")
        return
    
    print("\nBuscando los 10 archivos PDF/CSV más grandes...")
    largest_files = get_largest_files(directory)
    
    if not largest_files:
        print("No se encontraron archivos PDF o CSV en el directorio")
        return
    
    print("\nProcesando los archivos...")
    results = process_files(largest_files)
    
    # Mostrar resultados en tabla
    print("\nResultados:")
    print(tabulate(results, headers="keys", tablefmt="grid", showindex=True))

if __name__ == "__main__":
    # Instalar dependencias si no están disponibles
    try:
        from PyPDF2 import PdfReader
        from tabulate import tabulate
    except ImportError:
        print("Instalando dependencias necesarias...")
        import subprocess
        subprocess.run(['pip', 'install', 'PyPDF2', 'tabulate'], check=True)
        from PyPDF2 import PdfReader
        from tabulate import tabulate
    
    main()